r2/r2/lib/providers/search/solr.py

```diff
@@ -0,0 +1,847 @@
+    def __init__(self, solr_host=None, solr_port=None, fullnames=None):
+        self.solr_host = solr_host or g.solr_doc_host
+        self.solr_port = solr_port or g.solr_port
+        self.fullnames = fullnames    
+
+    @classmethod
+    def desired_fullnames(cls, items):
+        '''Pull fullnames that represent instances of 'types' out of items'''
+
+        fullnames = set()
+        type_ids = [type_._type_id for type_ in cls.types]
+        for item in items:
+            item_type = r2utils.decompose_fullname(item['fullname'])[1]
+            if item_type in type_ids:
+                fullnames.add(item['fullname'])
+        return fullnames
+
+    def add_xml(self, thing):
+
+        doc = etree.Element("doc")
+        field = etree.SubElement(doc, "field", name='id')
+        field.text = thing._fullname
+
+        for field_name, value in self.fields(thing).iteritems():
+            field = etree.SubElement(doc, "field", name=field_name)
+            field.text = safe_xml_str(value)
+
+        return doc
+
+    def delete_xml(self, thing):
+        '''Return the solr XML representation of
+        "delete this from the index"
+        
+        '''
+        delete = etree.fromstring('<id>%s</id>' % thing._id)
+        return delete
+
+    def delete_ids(self, ids):
+        '''Delete documents from the index.
+        'ids' should be a list of fullnames
+        
+        '''
+        deletes = [etree.fromstring('<id>%s</id>' % id_) \
+                for id_ in ids]
+
+
+        batch = etree.Element("delete")
+        batch.extend(deletes)
+        return self.send_documents(batch)
+
+    def batch_lookups(self):                                                                                            
+        try:                                                                                                            
+            self.things = Thing._by_fullname(self.fullnames, data=True,                                                 
+                                             return_dict=False)                                                         
+        except NotFound:                                                                                                
+            if self.use_safe_get:                                                                                       
+                self.things = safe_get(Thing._by_fullname, self.fullnames,                                              
+                                       data=True, return_dict=False)                                                    
+            else:                                                                                                       
+                raise 
+
+    def xml_from_things(self):
+        '''Generate a <batch> XML tree to send to solr for
+        adding/updating/deleting the given things
+        
+        '''
+        add = etree.Element("add")
+        delete = etree.Element("delete")
```

Comment:
In my testing, I found that the index wasn't ever getting updated with the uploaded results. I determined that the `<commit>` directive isn't ever sent. My recommendation would be to update this function to:

* Set `commit = etree.Element("commit"); commit.attribs["waitSearcher"] = "false"`
* `return (add, delete, commit)` from this function
* Update the loop in `inject()` to not skip an empty `commit` directive (since it will always be a child-less `etree.Element`)

https://cwiki.apache.org/confluence/display/solr/Uploading+Data+with+Index+Handlers

With the above changes, I was able to upload and search locally!